upgrade to wgpu v25

ygdrasil-io opened the pull request:
Upgrade to wgpu v25 and update the Android NDK version to v28.
I built a version for testing :
https://github.com/ygdrasil-io/wgpu-native/releases/tag/v25.0.0.1

ygdrasil-io committed:
Update to wgpu v25.0.1 and refactor for compatibility

Updated dependencies to wgpu-core, wgpu-types, wgpu-hal, and naga v25.0.1. Applied significant code changes to align with the updated wgpu API, including refactoring device polling, shader module descriptors, and unused features removal. Addressed compatibility issues and added necessary comments for unsupported features.

ygdrasil-io committed:
Remove unnecessary js-sys dependency from Cargo.toml

The js-sys dependency was not used in the project and is unnecessary. Cleaning up unused dependencies simplifies the configuration and reduces potential security risks.

ygdrasil-io committed:
Update Android NDK version in CI workflow file

Bumped the WGPU_ANDROID_NDK_VERSION from r27b to r28b to use the latest available version. This ensures the build process stays up-to-date with recent updates and fixes in the Android NDK. No other changes were made to the CI configuration.

ygdrasil-io committed:
run cargo fmt

ygdrasil-io committed:
Add Foundation framework to macOS build configuration

Updated CMakeLists.txt files to include the Foundation framework alongside CoreFoundation, QuartzCore, and Metal for macOS. This ensures proper linking and resolves potential dependency issues in macOS builds.

cwfitzgerald commented on src/conv.rs:
```diff
@@ -296,6 +296,7 @@ pub unsafe fn map_instance_descriptor(
                 (Some(dxil_path), Some(dxc_path)) => wgt::Dx12Compiler::DynamicDxc {
                     dxil_path: dxil_path.to_string(),
                     dxc_path: dxc_path.to_string(),
+                    max_shader_model: wgt::DxcShaderModel::V6_7,
```
We should add this as a member in `extras`

ygdrasil-io replied:
fixed

cwfitzgerald commented on src/conv.rs:
```diff
@@ -307,10 +308,12 @@ pub unsafe fn map_instance_descriptor(
             backend_options: wgt::BackendOptions {
                 gl: wgt::GlBackendOptions {
                     gles_minor_version: map_gles3_minor_version(extras.gles3MinorVersion),
+                    fence_behavior: Default::default(),
```
This as well

ygdrasil-io replied:
fixed

cwfitzgerald commented on src/conv.rs:
```diff
@@ -1142,11 +1144,12 @@ pub fn features_to_native(features: wgt::Features) -> Vec<native::WGPUFeatureNam
-    {
-        temp.push(native::WGPUNativeFeature_UniformBufferAndStorageTextureArrayNonUniformIndexing);
-    }
+    // TODO: fix this, UNIFORM_BUFFER_AND_STORAGE_TEXTURE_ARRAY_NON_UNIFORM_INDEXING is not supported anymore https://github.com/gfx-rs/wgpu/issues/4407
+    // if features
+    //     .contains(wgt::Features::UNIFORM_BUFFER_AND_STORAGE_TEXTURE_ARRAY_NON_UNIFORM_INDEXING)
+    // {
+    //     temp.push(native::WGPUNativeFeature_UniformBufferAndStorageTextureArrayNonUniformIndexing);
+    // }
```
Could you file an issue about fixing these, or include the fix in the PR?

almarklein replied:
We can simply remove it from `wgpu.h`.

ygdrasil-io replied:
I think we should review all the enumerations at once, as some are commented. Is there already an issue for tacking this?

almarklein replied:
Yes, would be good to review the feature enums 👌   Issue #190 seems related.

ygdrasil-io replied:
If that's alright for everybody, I'm keeping it as is, as this will be fixed in #190.

cwfitzgerald commented on src/lib.rs:
```diff
@@ -4282,14 +4279,15 @@ pub unsafe extern "C" fn wgpuDevicePoll(
 
     match context.device_poll(device_id, maintain) {
-        Ok(queue_empty) => queue_empty,
+        // TODO check if we need to check the PollStatus to return true or false
+        Ok(_) => true,
```
yes, this needs to match if PollStatus == QueueEmpty

ygdrasil-io replied:
Should we log something when it does not match ?

cwfitzgerald replied:
I mean, `matches!(poll_status, QueueEmpty)` instead of always true

ygdrasil-io replied:
I get it, but should we add a log to explain what is happening when we return false ?

ygdrasil-io replied:
fixed

cwfitzgerald reviewed (changes requested):
Looking good, some concerns

cwfitzgerald commented:
With the new behavior in features, I think we can use a single unconditional dep on wgc with the dx12, gles, vulkan, and metal features.

ygdrasil-io committed:
Fix device poll logic to handle PollStatus correctly

Updated the match statement to return true only when PollStatus is QueueEmpty. This clarifies the logic and ensures accurate handling of the polling status.

ygdrasil-io committed:
Add support for mapping DirectX max shader model versions

Introduce `map_dxc_max_shader_model` to handle DirectX shader model mappings, ensuring extensibility for additional versions. Updated related structures and enums in both Rust and C headers to accommodate `dxcMaxShaderModel`. This replaces the hardcoded shader model value in the pipeline descriptor logic.

ygdrasil-io committed:
Add support for GL fence behavior mapping

Introduce `WGPUGLFenceBehaviour` enum and add mapping for GL fence behaviors via `map_gl_fence_behavior`. This enhances OpenGL backend configuration by enabling proper handling of fence behavior settings in the instance descriptor.

ygdrasil-io committed:
Fix missing semicolon in WGPUInstanceExtras struct

A semicolon was missing after the `dxcMaxShaderModel` member, which could cause compilation errors. This change ensures the struct definition is syntactically correct.

ygdrasil-io commented:
@cwfitzgerald are you expecting more changes ?

cwfitzgerald commented:
Just didn't realize this was updated. I'll see if I can take a look at it soon-ish

cwfitzgerald commented on ffi/wgpu.h:
```diff
@@ -118,14 +118,34 @@ typedef enum WGPUNativeQueryType {
     WGPUNativeQueryType_Force32 = 0x7FFFFFFF
 } WGPUNativeQueryType WGPU_ENUM_ATTRIBUTE;
 
+typedef enum WGPUDirectXMaxShaderModel {
```
I'm not sure we should deviate from `wgpu`'s name here, without good reason - this is specifically about the max shader model of the DXC exe the user is handing to wgpu. I think this should either be `WGPUDx12ShaderModel` or `WGPUDxcMaxShaderModel`

ygdrasil-io replied:
fixed

cwfitzgerald reviewed (approved):
Good to go after one comment and a rebase.

ygdrasil-io committed:
Rename WGPUDirectXMaxShaderModel as WGPUDxcMaxShaderModel

ygdrasil-io committed:
Update version to v25.0.2

ygdrasil-io commented:
- Rebase done
- WGPUDirectXMaxShaderModel changed to WGPUDxcMaxShaderModel
- Upgraded to latest minor version

@cwfitzgerald or @almarklein If that's all right, can you merge and do a release please?

almarklein committed:
Merge branch 'trunk' into feature/upgrade-to-wgpu-v25

almarklein commented:
Thanks! 🚀
